File: app/src/main/java/com/saitejajanjirala/imdbclone/adapters/HomePageAdapters.java
package com.saitejajanjirala.imdbclone.adapters;

import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.saitejajanjirala.imdbclone.R;

import com.saitejajanjirala.imdbclone.databinding.BookMarkedItemBinding;
import com.saitejajanjirala.imdbclone.databinding.HomePageItemBinding;
import com.saitejajanjirala.imdbclone.models.Result;
import com.saitejajanjirala.imdbclone.listeners.BookMarkListener;
import com.saitejajanjirala.imdbclone.utils.Utils;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class HomePageAdapters extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements BookMarkListener {
    private List<Result> results;
    private Context context;
    private onHomePageItemClickListener listener;
    private int type;


    public HomePageAdapters(List<Result> results, Context context, onHomePageItemClickListener listener, int type) {
        this.results = results;
        this.context = context;
        this.listener = listener;
        this.type = type;
    }


    @NonNull
    @NotNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
        if(viewType==0){
            HomePageItemBinding binding=HomePageItemBinding.inflate(LayoutInflater.from(parent.getContext()),parent,false);
            return new HomePageViewHolder(binding);
        }
        else{
            Log.i("bookmark viewholder","bookmark");
            BookMarkedItemBinding binding=BookMarkedItemBinding.inflate(LayoutInflater.from(parent.getContext()),parent,false);
            return new BookMarkViewHolder(binding);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull @NotNull RecyclerView.ViewHolder holder, int position) {
        if(position!=RecyclerView.NO_POSITION) {
            if(holder instanceof HomePageViewHolder) {
                Result result = results.get(position);
                ((HomePageViewHolder) holder).bindData(result);
            }
            else if(holder instanceof BookMarkViewHolder){
                Log.i("bookmark binding","called");
                Result result = results.get(position);
                ((BookMarkViewHolder) holder).bindData(result);
            }
        }
    }



    @Override
    public int getItemCount() {
        return results.size();
    }

    @Override
    public void onBookMarkClicked(int position, Result result) {
        this.results.set(position, result);
        notifyItemChanged(position);
    }

    public class HomePageViewHolder extends RecyclerView.ViewHolder {
        HomePageItemBinding binding;
        public HomePageViewHolder(@NonNull @NotNull HomePageItemBinding binding) {
            super(binding.getRoot());
            this.binding=binding;
            binding.imdbPoster.setOnClickListener(v -> {
                int position=HomePageViewHolder.this.getAdapterPosition();
                if(position!=RecyclerView.NO_POSITION){
                    if(listener!=null){
                        listener.onItemClick(results.get(position));
                    }
                }
            });
            binding.bookMark.setOnClickListener(v->{
                int position=HomePageViewHolder.this.getAdapterPosition();
                if(position!=RecyclerView.NO_POSITION){
                    if(listener!=null){
                        listener.onBookMarkClicked(results.get(position),position,HomePageAdapters.this);
                    }
                }
            });
        }

        void bindData(Result result) {
            Utils.LoadImage(context,result.poster_path,binding.imdbPoster);
            if (result.adult) {
                binding.eighteenPlus.setVisibility(View.VISIBLE);
            }
            binding.title.setText(result.title);
            if(result.release_date!=null) {
                binding.year.setText(result.release_date.split("-")[0]);
            }
            else{
                binding.year.setText("-");
            }
            if(result.bookmarked){
                binding.bookMark.setImageDrawable(
                        ContextCompat.getDrawable(context,R.drawable.bookmark)
                );
            }
            else{
                binding.bookMark.setImageDrawable(
                        ContextCompat.getDrawable(context,R.drawable.un_bookmark)
                );
            }
            binding.rating.setText(result.vote_average+"");
        }
    }
    public class BookMarkViewHolder extends RecyclerView.ViewHolder {
        BookMarkedItemBinding binding;
        public BookMarkViewHolder(@NonNull @NotNull BookMarkedItemBinding binding) {
            super(binding.getRoot());
            this.binding=binding;
            binding.imdbPoster.setOnClickListener(v -> {
                int position=BookMarkViewHolder.this.getAdapterPosition();
                if(position!=RecyclerView.NO_POSITION){
                    if(listener!=null){
                        listener.onItemClick(results.get(position));
                    }
                }
            });
            binding.bookMark.setOnClickListener(v->{
                int position=BookMarkViewHolder.this.getAdapterPosition();
                if(position!=RecyclerView.NO_POSITION){
                    if(listener!=null){
                        listener.onBookMarkClicked(results.get(position),position,HomePageAdapters.this);
                    }
                }
            });
        }

        void bindData(Result result) {
            Utils.LoadImage(context,result.poster_path,binding.imdbPoster);
            if (result.adult) {
                binding.eighteenPlus.setVisibility(View.VISIBLE);
            }
            binding.title.setText(result.title);
            if(result.release_date!=null) {
                binding.year.setText(result.release_date.split("-")[0]);
            }
            else{
                binding.year.setText("-");
            }
            if(result.bookmarked){
                binding.bookMark.setImageDrawable(
                        ContextCompat.getDrawable(context,R.drawable.bookmark)
                );
            }
            else{
                binding.bookMark.setImageDrawable(
                        ContextCompat.getDrawable(context,R.drawable.un_bookmark)
                );
            }
            binding.rating.setText(result.vote_average+"");
        }
    }


    public interface onHomePageItemClickListener{
        void onItemClick(Result result);
        void onBookMarkClicked(Result result,int position,BookMarkListener listener);
    }

    public int getType() {
        return type;
    }
}

File: app/src/main/java/com/saitejajanjirala/imdbclonekotlin/adapters/HomePageAdapters.kt
package com.saitejajanjirala.imdbclonekotlin.adapters

import android.content.Context
import android.util.Log
import androidx.recyclerview.widget.RecyclerView
import com.saitejajanjirala.imdbclonekotlin.listeners.BookMarkListener
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import androidx.core.content.ContextCompat
import com.saitejajanjirala.imdbclonekotlin.R
import com.saitejajanjirala.imdbclonekotlin.databinding.BookMarkedItemBinding
import com.saitejajanjirala.imdbclonekotlin.databinding.HomePageItemBinding
import com.saitejajanjirala.imdbclonekotlin.utils.Helper
import com.saitejajanjirala.imdbclonekotlin.models.Result
class HomePageAdapters(
    private val results: MutableList<Result>,
    private val context: Context,
    private val listener: onHomePageItemClickListener?,
    val type: Int
) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), BookMarkListener {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == 0) {
            val binding: HomePageItemBinding =
                HomePageItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            HomePageViewHolder(binding)
        } else {
            Log.i("bookmark viewholder", "bookmark")
            val binding: BookMarkedItemBinding =
                BookMarkedItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            BookMarkViewHolder(binding)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (position != RecyclerView.NO_POSITION) {
            if (holder is HomePageViewHolder) {
                val result = results[position]
                holder.bindData(result!!)
            } else if (holder is BookMarkViewHolder) {
                Log.i("bookmark binding", "called")
                val result = results[position]
                holder.bindData(result!!)
            }
        }
    }

    override fun getItemCount(): Int {
        return results.size
    }

    override fun onBookMarkClicked(position: Int, result: Result?) {
        results[position] = result!!
        notifyItemChanged(position)
    }

    inner class HomePageViewHolder(binding: HomePageItemBinding) :
        RecyclerView.ViewHolder(binding.root) {
        var binding: HomePageItemBinding
        fun bindData(result: Result) {
            result.poster_path?.let { Helper.loadImage(context, it, binding.imdbPoster) }
            if (result.adult) {
                binding.eighteenPlus.visibility = View.VISIBLE
            }
            binding.title.text = result.title
            if (result.release_date != null) {
                binding.year.text = "${result.release_date}"
            } else {
                binding.year.text = "-"
            }
            if (result.bookmarked) {
                binding.bookMark.setImageDrawable(
                    ContextCompat.getDrawable(context, R.drawable.bookmark)
                )
            } else {
                binding.bookMark.setImageDrawable(
                    ContextCompat.getDrawable(context, R.drawable.un_bookmark)
                )
            }
            binding.rating.text = result.vote_average.toString() + ""
        }

        init {
            this.binding = binding
            binding.imdbPoster.setOnClickListener { v ->
                val position = this@HomePageViewHolder.adapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    listener?.onItemClick(results[position])
                }
            }
            binding.bookMark.setOnClickListener { v ->
                val position = this@HomePageViewHolder.adapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    listener?.onBookMarkClicked(results[position], position, this@HomePageAdapters)
                }
            }
        }
    }

    inner class BookMarkViewHolder(binding: BookMarkedItemBinding) :
        RecyclerView.ViewHolder(binding.root) {
        var binding: BookMarkedItemBinding = binding
        fun bindData(result: Result) {
            result.poster_path?.let { Helper.loadImage(context, it, binding.imdbPoster) }
            if (result.adult) {
                binding.eighteenPlus.visibility = View.VISIBLE
            }
            binding.title.text = result.title
            if (result.release_date != null) {
                binding.year.text = "${result.release_date}"
            } else {
                binding.year.text = "-"
            }
            if (result.bookmarked) {
                binding.bookMark.setImageDrawable(
                    ContextCompat.getDrawable(context, R.drawable.bookmark)
                )
            } else {
                binding.bookMark.setImageDrawable(
                    ContextCompat.getDrawable(context, R.drawable.un_bookmark)
                )
            }
            binding.rating.text = result.vote_average.toString() + ""
        }

        init {
            binding.imdbPoster.setOnClickListener { v ->
                val position = this@BookMarkViewHolder.adapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    listener?.onItemClick(results[position])
                }
            }
            binding.bookMark.setOnClickListener { v ->
                val position = this@BookMarkViewHolder.adapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    listener?.onBookMarkClicked(results[position], position, this@HomePageAdapters)
                }
            }
        }
    }

    interface onHomePageItemClickListener {
        fun onItemClick(result: Result?)
        fun onBookMarkClicked(result: Result?, position: Int, listener: BookMarkListener?)
    }
}